File: Assignment1/4ProcessData/clean_csv.java
import java.io.*;
import java.util.ArrayList;
import java.util.List;



public class clean_csv {
    public static void main(String[] args) throws IOException {
        List<String> a_line=new ArrayList<>();
        List<String> final_line=new ArrayList<>();
        File csv = new File("C:\\Users\\Lenovo\\Desktop\\datawarehouse_productData.csv");  // CSV文件路径
        File outFile = new File("C:\\Users\\Lenovo\\Desktop\\out.csv");
        BufferedWriter writer = new BufferedWriter(new FileWriter(outFile));
        BufferedReader br = null;
        try
        {
            br = new BufferedReader(new FileReader(csv));
        } catch (FileNotFoundException e)
        {
            e.printStackTrace();
        }
        String line = "";
        String everyLine = "";
        try {
            List<String> allString = new ArrayList<>();
            while ((line = br.readLine()) != null)  //读取到的内容给line变量
            {
                everyLine = line;

                //System.out.println(everyLine);
                allString.add(everyLine);
            }
            a_line=allString;
            //System.out.println("csv表格中所有行数："+allString.size());
        } catch (IOException e)
        {
            e.printStackTrace();
        }
        String temp1=a_line.get(0);
        for (int i = 0; i < a_line.size(); i++) {
            System.out.println("第"+i+"轮");
            String temp2=a_line.get(i);

            String[] temp1_list=temp1.split(",");
            System.out.println(temp1_list[1]);

            String[] temp2_list=temp2.split(",");
            System.out.println(temp2_list[1]);

            if(levenshtein(temp1_list[1],temp2_list[1])>0.4)
                continue;
            final_line.add(temp1);
            temp1=temp2;
        }
        final_line.add(temp1);
        for (int i = 0; i < final_line.size(); i++) {
            System.out.println(final_line.get(i));
        }
        for (int i = 0; i < final_line.size(); i++) {
            System.out.println("第"+i+"轮");
            writer.write(final_line.get(i));
            writer.newLine();
        }
        System.out.println(final_line.size());
    }

    //相似度匹配算法
    public static float levenshtein(String str1,String str2) {
        //计算两个字符串的长度。
        int len1 = str1.length();
        int len2 = str2.length();
        //建立上面说的数组，比字符长度大一个空间
        int[][] dif = new int[len1 + 1][len2 + 1];
        //赋初值，步骤B。
        for (int a = 0; a <= len1; a++) {
            dif[a][0] = a;
        }
        for (int a = 0; a <= len2; a++) {
            dif[0][a] = a;
        }
        //计算两个字符是否一样，计算左上的值
        int temp;
        for (int i = 1; i <= len1; i++) {
            for (int j = 1; j <= len2; j++) {
                if (str1.charAt(i - 1) == str2.charAt(j - 1)) {
                    temp = 0;
                } else {
                    temp = 1;
                }
                //取三个值中最小的
                dif[i][j] = min(dif[i - 1][j - 1] + temp, dif[i][j - 1] + 1,
                        dif[i - 1][j] + 1);
            }
        }
        //System.out.println("字符串\""+str1+"\"与\""+str2+"\"的比较");
        //取数组右下角的值，同样不同位置代表不同字符串的比较
        //System.out.println("差异步骤："+dif[len1][len2]);
        //计算相似度
        float similarity =1 - (float) dif[len1][len2] / Math.max(str1.length(), str2.length());
        //System.out.println("相似度："+similarity);
        return similarity;
    }

    //得到最小值
    private static int min(int... is) {
        int min = Integer.MAX_VALUE;
        for (int i : is) {
            if (min > i) {
                min = i;
            }
        }
        return min;
    }
}

File: Assignment1/4ProcessData/clean_csv.py
import csv
import difflib


def string_similar(s1, s2):
    return difflib.SequenceMatcher(None, s1, s2).ratio()


def Max(s1, s2):
    if(len(s1)>len(s2)):
        return s1
    else:
        return s2

def JudgeTime(s1,s2):
    if s1==s2:
        return True
    elif s1 == '':
        return True
    elif s2 == '':
        return True
    else:
        return False
def Similarity(s1, s2):
    len1 = len(s1)
    len2 = len(s2)
    dif = [[]]
    for a in range(len1+2):
        dif[a][0] = a
    for a in range(len2+2):
        dif[0][a] = a
    for i in range(1,len1+1):
        for j in range(1,len2+1):
            if s1[i-1] == s2[j-1]:
                temp = 0
            else:
                temp = 1
            dif[i][j] = min(dif[i - 1][j - 1] + temp, dif[i][j - 1] + 1, dif[i - 1][j] + 1)
    similarity = 1 - dif[len1][len2] / max(len(s1), len(s2))
    return similarity


if __name__ == '__main__':
    with open('C:\\Users\\Lenovo\\Desktop\\productData.csv','r')as f:
        reader = csv.DictReader(f, fieldnames=('id','title','release date','genres','director','producers','actor','supporting actors','media format','run time','MPAA rating','subtitles','studio','Item model number','Date First Available','IMDb','audio languages'))
        final=[]
        out_final=[]
        for i in reader:
            final.append(dict(i))
        with open('C:\\Users\\Lenovo\\Desktop\\final_out1.csv', 'w', newline='')as outFile:
            writer = csv.DictWriter(outFile,  fieldnames=('id','title','release date','genres','director','producers','actor','supporting actors','media format','run time','MPAA rating','subtitles','studio','Item model number','Date First Available','IMDb','audio languages','link id','link title'))
            # writer.writerows(final)
            temp1 = final[0]
            temp1['link id'] = ''
            temp1['link title'] = ''
            for i in range(len(final)):
                temp2 = final[i]
                temp2['link id'] = ''
                temp2['link title'] = ''

                min_lenth = min(len(temp1['title']), len(temp2['title']))
                min_lenth = min(min_lenth, 12)
                min_lenth2 = min(min_lenth, 6)
                # if temp1['title'] == temp2['title'] and len(temp1['title']) <= 4 and len(temp2['title']) <= 4:
                #     temp1['director'] = Max(temp1['director'], temp2['director'])
                #     temp1['actor'] = Max(temp1['actor'], temp2['actor'])
                #     temp1['supporting actors'] = Max(temp1['supporting actors'], temp2['supporting actors'])
                #     temp1['genres'] = Max(temp1['genres'], temp2['genres'])
                #     temp1['producers'] = Max(temp1['producers'], temp2['producers'])
                #     temp1['run time'] = Max(temp1['run time'], temp2['run time'])
                #     temp1['Date First Available'] = Max(temp1['Date First Available'], temp2['Date First Available'])
                #     # temp1['director'] = Max(temp1['director'], temp2['director'])
                #     temp1['link id'] = temp1['link id'] + "\n" + temp2['id']
                #     temp1['link title'] = temp1['link title'] + "\n" + temp2['title']
                #     continue
                # elif string_similar(temp1['title'], temp2['title']) > 0.6 and JudgeTime(temp1['release date'],temp2['release date']):
                #     temp1['director'] = Max(temp1['director'], temp2['director'])
                #     temp1['actor'] = Max(temp1['actor'], temp2['actor'])
                #     temp1['supporting actors'] = Max(temp1['supporting actors'], temp2['supporting actors'])
                #     temp1['genres'] = Max(temp1['genres'], temp2['genres'])
                #     temp1['producers'] = Max(temp1['producers'], temp2['producers'])
                #     temp1['run time'] = Max(temp1['run time'], temp2['run time'])
                #     temp1['Date First Available'] = Max(temp1['Date First Available'], temp2['Date First Available'])
                #     # temp1['director'] = Max(temp1['director'], temp2['director'])
                #     temp1['link id'] = temp1['link id'] + "\n" + temp2['id']
                #     temp1['link title'] = temp1['link title'] + "\n" + temp2['title']
                #     continue
                # elif string_similar(temp1['title'][:min_lenth], temp2['title'][:min_lenth]) > 0.9 and len(temp1['title']) >= 8 and len(temp2['title']) >= 8 and len(temp1['title']) <= 18 and len(temp2['title']) <= 18:
                #     temp1['director'] = Max(temp1['director'], temp2['director'])
                #     temp1['actor'] = Max(temp1['actor'], temp2['actor'])
                #     temp1['supporting actors'] = Max(temp1['supporting actors'], temp2['supporting actors'])
                #     temp1['genres'] = Max(temp1['genres'], temp2['genres'])
                #     temp1['producers'] = Max(temp1['producers'], temp2['producers'])
                #     temp1['run time'] = Max(temp1['run time'], temp2['run time'])
                #     temp1['Date First Available'] = Max(temp1['Date First Available'], temp2['Date First Available'])
                #     # temp1['director'] = Max(temp1['director'], temp2['director'])
                #     temp1['link id'] = temp1['link id'] + "\n" + temp2['id']
                #     temp1['link title'] = temp1['link title'] + "\n" + temp2['title']
                #     continue
                # elif string_similar(temp1['title'][:min_lenth2], temp2['title'][:min_lenth2]) > 0.9 and len(temp1['title']) >= 4 and len(temp2['title']) >= 4:
                #     temp1['director'] = Max(temp1['director'], temp2['director'])
                #     temp1['actor'] = Max(temp1['actor'], temp2['actor'])
                #     temp1['supporting actors'] = Max(temp1['supporting actors'], temp2['supporting actors'])
                #     temp1['genres'] = Max(temp1['genres'], temp2['genres'])
                #     temp1['producers'] = Max(temp1['producers'], temp2['producers'])
                #     temp1['run time'] = Max(temp1['run time'], temp2['run time'])
                #     temp1['Date First Available'] = Max(temp1['Date First Available'], temp2['Date First Available'])
                #     # temp1['director'] = Max(temp1['director'], temp2['director'])
                #     temp1['link id'] = temp1['link id'] + "\n" + temp2['id']
                #     temp1['link title'] = temp1['link title'] + "\n" + temp2['title']
                #     continue

                if temp1['title'] == temp2['title'] and len(temp1['title']) <= 4 and len(temp2['title']) <= 4 and JudgeTime(temp1['release date'],temp2['release date']):
                    temp1['director'] = Max(temp1['director'], temp2['director'])
                    temp1['actor'] = Max(temp1['actor'], temp2['actor'])
                    temp1['supporting actors'] = Max(temp1['supporting actors'], temp2['supporting actors'])
                    temp1['genres'] = Max(temp1['genres'], temp2['genres'])
                    temp1['producers'] = Max(temp1['producers'], temp2['producers'])
                    temp1['run time'] = Max(temp1['run time'], temp2['run time'])
                    temp1['Date First Available'] = Max(temp1['Date First Available'], temp2['Date First Available'])
                    # temp1['director'] = Max(temp1['director'], temp2['director'])
                    temp1['link id'] = temp1['link id'] + "\n" + temp2['id']
                    temp1['link title'] = temp1['link title'] + "\n" + temp2['title']
                    continue
                elif string_similar(temp1['title'], temp2['title']) > 0.6 and JudgeTime(temp1['release date'],temp2['release date']):
                    temp1['director'] = Max(temp1['director'], temp2['director'])
                    temp1['actor'] = Max(temp1['actor'], temp2['actor'])
                    temp1['supporting actors'] = Max(temp1['supporting actors'], temp2['supporting actors'])
                    temp1['genres'] = Max(temp1['genres'], temp2['genres'])
                    temp1['producers'] = Max(temp1['producers'], temp2['producers'])
                    temp1['run time'] = Max(temp1['run time'], temp2['run time'])
                    temp1['Date First Available'] = Max(temp1['Date First Available'], temp2['Date First Available'])
                    # temp1['director'] = Max(temp1['director'], temp2['director'])
                    temp1['link id'] = temp1['link id'] + "\n" + temp2['id']
                    temp1['link title'] = temp1['link title'] + "\n" + temp2['title']
                    continue
                elif string_similar(temp1['title'][:min_lenth], temp2['title'][:min_lenth]) > 0.9 and len(temp1['title']) >= 8 and len(temp2['title']) >= 8 and len(temp1['title']) <= 18 and len(temp2['title']) <= 18 and JudgeTime(temp1['release date'],temp2['release date']):
                    temp1['director'] = Max(temp1['director'], temp2['director'])
                    temp1['actor'] = Max(temp1['actor'], temp2['actor'])
                    temp1['supporting actors'] = Max(temp1['supporting actors'], temp2['supporting actors'])
                    temp1['genres'] = Max(temp1['genres'], temp2['genres'])
                    temp1['producers'] = Max(temp1['producers'], temp2['producers'])
                    temp1['run time'] = Max(temp1['run time'], temp2['run time'])
                    temp1['Date First Available'] = Max(temp1['Date First Available'], temp2['Date First Available'])
                    # temp1['director'] = Max(temp1['director'], temp2['director'])
                    temp1['link id'] = temp1['link id'] + "\n" + temp2['id']
                    temp1['link title'] = temp1['link title'] + "\n" + temp2['title']
                    continue
                elif string_similar(temp1['title'][:min_lenth2], temp2['title'][:min_lenth2]) > 0.9 and len(temp1['title']) >= 4 and len(temp2['title']) >= 4 and JudgeTime(temp1['release date'],temp2['release date']):
                    temp1['director'] = Max(temp1['director'], temp2['director'])
                    temp1['actor'] = Max(temp1['actor'], temp2['actor'])
                    temp1['supporting actors'] = Max(temp1['supporting actors'], temp2['supporting actors'])
                    temp1['genres'] = Max(temp1['genres'], temp2['genres'])
                    temp1['producers'] = Max(temp1['producers'], temp2['producers'])
                    temp1['run time'] = Max(temp1['run time'], temp2['run time'])
                    temp1['Date First Available'] = Max(temp1['Date First Available'], temp2['Date First Available'])
                    # temp1['director'] = Max(temp1['director'], temp2['director'])
                    temp1['link id'] = temp1['link id'] + "\n" + temp2['id']
                    temp1['link title'] = temp1['link title'] + "\n" + temp2['title']
                    continue
                writer.writerow(temp1)
                # out_final.append(dict(temp1))
                temp1 = temp2
            writer.writerow(temp1)
            # writer.writerows(out_final)



        # for i in final:
        #     print(i)


